dataflow/flex-templates/pipeline_with_dependencies/requirements.txt

```diff
@@ -0,0 +1,320 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# This requirements.txt file lists all Python dependenciences that are installed
+# in Docker images in this example.
+
+# See the "Update the dependencies in requirements.txt" section in README.md
+
+#
+# This file is autogenerated by pip-compile with Python 3.11
```

Comment:
This worries me a little bit since maintaining such a huge list of all the direct and indirect dependencies with all pinned versions can be a hassle, and is confusing to understand.

Can we keep this manually generated with just the minimal requirements and just pin those versions? We can then let pip decide the versions for all the indirect dependencies.

Reply:
By the way, does a bot in python-docs-samples repo attempt to update dependencies in this example? If so, I think excluding this example from the bot scope might be better to illustrate the idea of the example.

Reply:
> We do have a bot to update dependencies, but I think they update them one at a time, not all at once.

there is a `pip check` step, so an incompatibility would be detected; in this case requirements can be regenerated using the instructions in the requirements.txt file.

> Do you think we could just include a note both in the README and in the requirements.txt file explaining something like "Beam recommends reproducible environments for ... reasons" and then provide a link to the docs with more details?

Here are the references: i'll take a look once again to see if it needs more info:

https://github.com/GoogleCloudPlatform/python-docs-samples/pull/11252/files#diff-9f37f333f1a7f503b7420877c8f1807848667931e53df73af3c97e0ef76716c5R19

https://github.com/GoogleCloudPlatform/python-docs-samples/pull/11252/files#diff-3257e9c42f5a156e4daa91bf25275715e3939279509a1b8d8f9654d12369d40bR141

https://github.com/GoogleCloudPlatform/python-docs-samples/pull/11252/files#diff-3257e9c42f5a156e4daa91bf25275715e3939279509a1b8d8f9654d12369d40bR169

Reply:
is it possible to disable dependency upgrade for this example? I worry that requirements.txt will be updated but dockerfile will not be updated.

Reply:
for example when a tool decides to upgrade apache beam to the latest version.

Reply:
> Do you think we could just include a note both in the README and in the requirements.txt

done

Reply:
To exclude this sample from updates please add an update to [renovate.json](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/renovate.json) in your PR.

You can use exclude your sample directory entirely (see [ignorePaths](https://docs.renovatebot.com/configuration-options/#ignorepaths))

or you can exclude certain dependencies in your sample with something like this:
```
{
  "packageRules": [
    {
      "matchPackageNames": ["the-dependency-to-restrict"],
      "ignorePaths": [
        "**/some-directory-containing-dependency/**",
        "**/another-directory-with-dependency/**"
      ]
    }
  ]
}
```

Reply:
I didn't know we could group all dependencies into a single PR. I think configuring that would be the best solution.

Reply:
Thank, I excluded the example from renovatebot.

Reply:
As per our chat, here's what updating requirements for this sample means:

1. modify setup.py that sets the beam's dependency
2. modify Dockerfile and comments to refer to a consistent version of Beam
3. regenerate requirements.txt using pip-compile and the same python version as used in Dockerfile

Renovate bot would not be able to check for Dockerfile/Beam version consistency, and it wouldn't be able to regenerate the requirements.txt file, so it's okay to exclude.